[BEAM-7310] Add support of Confluent Schema Registry for KafkaIO

aromanenko-dev opened the pull request:
Add `withSchemaRegistry()` which allows to specify Confluent Schema Registry (CSR) url and `withKeySchemaSubject()/withValueSchemaSubject()` to set a schema subject in case if keys and/or values are serialised with Avro schema stored in CSR.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

aromanenko-dev commented:
Run JavaPortabilityApi PreCommit

aromanenko-dev commented:
Run SQL PreCommit

aromanenko-dev commented:
Run Python PreCommit

aromanenko-dev commented:
Run Python2_PVR_Flink PreCommit

aromanenko-dev commented:
Run Java PreCommit

aromanenko-dev commented:
Run CommunityMetrics PreCommit

aromanenko-dev commented:
Run CommunityMetrics PreCommit

aromanenko-dev commented:
@reuvenlax @iemejia could you take a look?

aromanenko-dev commented:
Run Java PreCommit

iemejia commented:
CC @rangadi in case you have some time/interest to take a look

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java:
```diff
@@ -155,6 +172,25 @@ public boolean start() throws IOException {
+    SchemaRegistryClient registryClient;
+    String schemaRegistryURL =
+        (String)
+            spec.getConsumerConfig().get(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG);
```
Can you make `AbstractKafkaAvroSerDeConfig` into `KafkaAvroDeserializerConfig` as below.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java:
```diff
@@ -155,6 +172,25 @@ public boolean start() throws IOException {
+    } else {
+      registryClient =
+          new CachedSchemaRegistryClient(
+              (String)
```
This is `schemaRegistryURL` no?

aromanenko-dev replied:
Yes, good catch, will be fixed.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -817,6 +892,27 @@ public void setValueDeserializer(String valueDeserializer) {
     }
 
+    private Schema fetchAvroSchema(String schemaRegistryUrl, String subject) {
+      SchemaRegistryClient registryClient;
```
Can you refactor the code of `KafkaUnboundedReader#getSchemaRegistryClient` into a method that you use here to avoid the repeated code to get a client.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -862,19 +958,56 @@ public void setValueDeserializer(String valueDeserializer) {
+      Coder<K> keyCoder;
+      if (avroKeySchema != null) {
+        keyCoder = (Coder<K>) AvroCoder.of(avroKeySchema);
+        checkArgument(
```
but probably that check should be `checkState` per https://github.com/google/guava/wiki/PreconditionsExplained

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -391,6 +451,84 @@ public static void addCountingAsserts(
+  public void testReadAvroKeyValuesWithSchemaRegistry() {
+    int numElements = 100;
+    String topic = "my_topic";
+    String keySubject = topic + "-key";
```
Rename to `keySchemaSubject` (same for `valueSchemaSubject`)

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -391,6 +451,84 @@ public static void addCountingAsserts(
+    private final String keySchemaSubject;
+    private final String valueSchemaSubject;
+
+    private MockSchemaRegistryClientFactoryFn(String keySubject, String valueSubject) {
```
`keySchemaSubject` and `valueSchemaSubject`

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -160,14 +169,36 @@
       OffsetResetStrategy offsetResetStrategy,
-      Map<String, Object> config) {
+      Map<String, Object> config,
+      boolean isAvro) {
```
maybe rename `useAvro` is a more clear name.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -356,6 +366,15 @@
     @Nullable
     abstract Map<String, Object> getOffsetConsumerConfig();
 
+    @Nullable
```
On the class header (don't know why I pointed it here)

aromanenko-dev commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -817,6 +892,27 @@ public void setValueDeserializer(String valueDeserializer) {
+      }
+
+      final Schema avroSchema = new Schema.Parser().parse(latestSchemaMetadata.getSchema());
+      checkArgument(avroSchema != null, "Avro schema can't be null");
```
I guess it's almost impossible but I'd prefer to leave this check "just in case" since we can't be sure on 100% and I don't see any disadvantages to have it here.

aromanenko-dev commented:
Run JavaPortabilityApi PreCommit

aromanenko-dev commented:
Run JavaPortabilityApi PreCommit

aromanenko-dev commented:
Run CommunityMetrics PreCommit

aromanenko-dev commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -586,6 +612,55 @@ public void setValueDeserializer(String valueDeserializer) {
+     * should be used together with {@link #withKeySchemaSubject(String)} and/or {@link
+     * #withValueSchemaSubject(String)}.
+     */
+    public Read<K, V> withSchemaRegistry(String schemaRegistryUrl) {
```
Good question. I think that all different types of authentication can be configured through CSR options and will be used by `CachedSchemaRegistryClient` if any:
https://docs.confluent.io/current/schema-registry/security/index.html

aromanenko-dev commented:
@iemejia I addressed your comments, PTAL

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -175,6 +175,30 @@
+ *   .apply(KafkaIO.<Long, GenericRecord>read()
+ *      .withBootstrapServers("broker_1:9092,broker_2:9092")
+ *      .withTopic("my_topic")
+ *      .withSchemaRegistry(""http://localhost:8081"") // Confluent Schema Registry URL
```
nit: can you make this by default a secure connection in the example e.g. `https`

aromanenko-dev replied:
I'd leave this example as simple as possible. All configuration options for CSR can be found in its documentation

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -913,6 +933,21 @@ private Schema fetchAvroSchema(String schemaRegistryUrl, String subject) {
+        String schemaRegistryUrl,
+        SerializableFunction<String, SchemaRegistryClient> schemaRegistryClientFactoryFn) {
+      SchemaRegistryClient registryClient;
+      if (schemaRegistryUrl == null) {
```
`checkArgument` better, not having a URL to get the client is an Exception condition.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -971,20 +1023,18 @@ private Schema fetchAvroSchema(String schemaRegistryUrl, String subject) {
-                + "key coder explicitly using withKeyDeserializerAndCoder()");
+            getKeyCoder() != null ? getKeyCoder() : inferCoder(coderRegistry, getKeyDeserializer());
       }
+      checkArgument(
```
`checkState` better, we should use `checkArgument` only for arguments.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -994,34 +1044,17 @@ private Schema fetchAvroSchema(String schemaRegistryUrl, String subject) {
-      }
-
-      return input.getPipeline().apply(transform);
+      checkArgument(
```
`checkState` better, we should use `checkArgument` only for arguments.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java:
```diff
@@ -91,22 +90,34 @@ public boolean start() throws IOException {
     try {
       if (spec.getKeyDeserializer().equals(KafkaAvroDeserializer.class)) {
-        SchemaRegistryClient registryClient = getSchemaRegistryClient(spec);
+        if (registryClient == null) {
```
Remove validation per above comment on checkArgument, if we cannot build a client we fail.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java:
```diff
@@ -91,22 +90,34 @@ public boolean start() throws IOException {
 
       if (spec.getValueDeserializer().equals(KafkaAvroDeserializer.class)) {
-        SchemaRegistryClient registryClient = getSchemaRegistryClient(spec);
+        if (registryClient == null) {
```
Remove too (see above)

rangadi commented:
Thanks for the PR @aromanenko-dev & @iemejia for thorough review. Overall LGTM. Just one thing to consider in naming: the doc comments always mention 'Confluent Schema Registry', probably implying this is specific to Confluent. Should the API names include `Confluent` or `CSR` in them too?

aromanenko-dev commented:
@iemejia I addressed your latest comments , PTAL

aromanenko-dev commented:
@rangadi Thank you for feedback! I agree with your naming proposal despite that afaik only one Avro Schema Registry exists (Confluent's one). In the same time, to avoid a confusion with Beam schema we should probably name it explicitly `withConfluentSchemaRegistry()`. Do you think we need to name other new public methods in this way, like `getKeySchemaSubject()` and `getValueSchemaSubject()` should be `getKeyConfluentSchemaSubject()` and `getValueConfluentSchemaSubject()` ? 

@rangadi @iemejia Also, now I'm thinking to reduce the number of new public methods, introduced by this PR, and to add only `withConfluentSchemaRegistry(ConfluentSchemaRegistryProvider)` where `ConfluentSchemaRegistryProvider` will incorporate all required members (registry URL, subjects, ClientFactoryFn) to use CSR in `KafkaIO`, like we do for KinesisIO with `AWSClientsProvider`. Wdyt?

iemejia commented:
Thanks for changes code looks ok for me :+1: 
About your proposals you have a valid point about Beam's schema in the naming (hard to reconcile): 
+1 for `withConfluentSchemaRegistry`
0 for getKeyConfluentSchemaSubject and getValueConfluentSchemaSubject 
The idea of the new object to have only one new signature sounds GORGEOUS to me please PLEASE !
Please tell me when can I take a look again, and thanks for your patience during the review.

reuvenlax commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -891,6 +1014,50 @@ public void setValueDeserializer(String valueDeserializer) {
+          "Value coder could not be inferred from value deserializer. Please provide"
+              + "value coder explicitly using withValueDeserializerAndCoder()");
+      return valueCoder;
+    }
```
It would be very useful to set a Beam schema based on the Avro schema instead of using AvroCoder. This would allow KafkaIO to be used as a source for Beam SQL

iemejia replied:
Excellent idea, I would suggest to use a more conservative approach to enable Beam Schema Coders similar to the one used for `set/getInferBeamSchema` in `AvroIO`.

aromanenko-dev replied:
I planned to do this in the next PR for https://issues.apache.org/jira/browse/BEAM-7336

aromanenko-dev commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -165,6 +175,30 @@
+ *   .apply(KafkaIO.<Long, GenericRecord>read()
+ *      .withBootstrapServers("broker_1:9092,broker_2:9092")
+ *      .withTopic("my_topic")
+ *      .withSchemaRegistry(""http://localhost:8081"") // Confluent Schema Registry URL
```
@reuvenlax In multiple topics case we expect that all topics follow the same schema structure which is stored in Confluent Schema Registry. This is like what we do with deserialisers (and coders) in such case - we expect that all messages from different topics can be deserialised by the same deserialiser. I was thinking to create a kind of map for topic->schema relations and deserialise it accordingly, but there are two objections against that:
- we will need to write a wrapper around `KafkaAvroDeserializer` to support multiple topics with multiple schemas there, which is probably not a big deal but having an output PCollection of `GenericRecord`s created based on different schemas seems quite complex.
- afaik, we can set only one single Beam schema for PCollection. So what we should do in multiple schemas case? 

As for now, I think we need to clear explain it in Javadoc and leave it as a user responsibility to choose a proper schema for all required input topic(s) that satisfy it.

Please, let me know if you have other ideas or examples about that, I'd be happy to discuss them to find a best solution.

aromanenko-dev commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -586,6 +636,55 @@ public void setValueDeserializer(String valueDeserializer) {
+     * serialisation. Also, sets {@link KafkaAvroDeserializer} as a key deserializer. {@link Coder}
+     * will be inferred based on obtained Avro schema by {@link AvroCoder}. It should be used along
+     * with {@link #withSchemaRegistry(String)}.
+     */
```
Please, see my comment above.

aromanenko-dev commented:
As we discussed before, I simplified KafkaIO API to use it with Confluent Schema Registry. I added a `CSRClientProvider` interface and two implementations - `BasicCSRClientProvider` (used by default in KafkaIO) and `TestCSRClientProvider` (for unit tests). The instance of provider should create and return the instance of `SchemaRegistryClient` which is used to fetch Avro schema from CSR.

Also, I added several another tests to test reading of key and values if they Avro serialised in different combinations.

I'm still on the subject if we need to support different schemas from different CSR subjects if user reads from multiple topics. Any ideas will be very welcomed.

aromanenko-dev commented:
Run Java KafkaIO Performance Test

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -586,6 +628,40 @@ public void setValueDeserializer(String valueDeserializer) {
+     * Registry. You should use this method if {@link Read#withCSRClientProvider(String, String,
+     * String)} does not suit your needs.
+     */
+    public Read<K, V> withCSRClientProvider(CSRClientProvider csrClientProvider) {
```
I would make this one package private until someone really asks for it.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/BasicCSRClientProvider.java:
```diff
@@ -0,0 +1,56 @@
+import javax.annotation.Nullable;
+
+/** Basic implementation of {@link CSRClientProvider} used by default in {@link KafkaIO}. */
+public class BasicCSRClientProvider implements CSRClientProvider {
```
package private, see my comment on `withCSRClientProvider` below

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -165,6 +173,35 @@
+ *
+ * <pre>{@code
+ * // Confluent Schema Registry URL
+ * String schemaRegistryUrl = "http://localhost:8081";
```
I would better let the documentation with the 3-arg method, prefer simpler use, we can easily evolve towards the more generic case if there is a need which so far seems not to be the case.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/BasicCSRClientProvider.java:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  @Override
+  public String getSchemaRegistryUrl() {
```
Remove, see comment on the interface below

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -586,6 +628,40 @@ public void setValueDeserializer(String valueDeserializer) {
+
+    /**
+     * Specify Confluent schema registry url, key and value schema subjects to be used to read
+     * schema from Confluent Schema Registry. If you need more sophisticated credential protocol,
```
Remove the comment `"If you need more sophisticated version..."` for consistency with non public exposure of the related method.

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -817,6 +893,20 @@ public void setValueDeserializer(String valueDeserializer) {
       return new TypedWithoutMetadata<>(this);
     }
 
+    private Schema fetchAvroSchema(CSRClientProvider csrClientProvider, String subject) {
```
static

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -817,6 +893,20 @@ public void setValueDeserializer(String valueDeserializer) {
+      }
+
+      final Schema avroSchema = new Schema.Parser().parse(latestSchemaMetadata.getSchema());
+      checkArgument(avroSchema != null, "Avro schema can't be null");
```
`checkState`

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaAvroGenericRecordExample.java:
```diff
@@ -0,0 +1,53 @@
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+public class KafkaAvroGenericRecordExample {
```
I have mixed feelings about this one with a tendency towards remove it. I like the idea, but at the same time is like more code to maintain (that is not tested). Are we doing this for other IOs or other parts of the codebase.

aromanenko-dev replied:
You are right, I'll remove it. Though, we will need to add this a part of KafkaIO IT later.

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -1784,4 +1966,75 @@ void shutdown() {
     }
   }
+
+  private static class AvroDeSe {
```
better call it `AvroDeSer` if it is a game of words around `AvroSerDe` or even better just call it `AvroSerDe`

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -1784,4 +1966,75 @@ void shutdown() {
+  }
+
+  private static class KeyAvroSerializableFunction extends BaseAvroSerializableFunction {
+    KeyAvroSerializableFunction(String topic, String schemaRegistryUrl, boolean isKey) {
```
remove `isKey` and fix it in the call to super to `true`

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -1784,4 +1966,75 @@ void shutdown() {
+  }
+
+  private static class ValueAvroSerializableFunction extends BaseAvroSerializableFunction {
+    ValueAvroSerializableFunction(String topic, String schemaRegistryUrl, boolean isKey) {
```
remove `isKey` and fix it in the call to super to `false`

iemejia commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -391,6 +427,152 @@ public static void addCountingAsserts(
+  }
+
+  @Test
+  public void testReadAvroValuesAsPojoRecords() {
```
`s/PojoRecord/SpecificRecords` by Avro common parlance

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -891,6 +970,52 @@ public void setValueDeserializer(String valueDeserializer) {
+        String schemaSubject = csrClientProvider.getKeySchemaSubject();
+        if (schemaSubject != null) {
+          avroKeySchema = fetchAvroSchema(csrClientProvider, schemaSubject);
+          checkArgument(avroKeySchema != null, "Avro key schema can't be null");
```
`checkState`

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -891,6 +970,52 @@ public void setValueDeserializer(String valueDeserializer) {
+        String schemaSubject = csrClientProvider.getValueSchemaSubject();
+        if (schemaSubject != null) {
+          avroValueSchema = fetchAvroSchema(csrClientProvider, schemaSubject);
+          checkArgument(avroValueSchema != null, "Avro value schema can't be null");
```
`checkState`

iemejia commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/TestCSRClientProvider.java:
```diff
@@ -0,0 +1,64 @@
+import org.apache.avro.Schema;
+
+/** Mock implementation of {@link CSRClientProvider} used for testing. */
+public class TestCSRClientProvider extends BasicCSRClientProvider {
```
Move this file better `src/test/java` and make the class package private

aromanenko-dev commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/CSRClientProvider.java:
```diff
@@ -0,0 +1,38 @@
+
+  SchemaRegistryClient getCSRClient();
+
+  String getSchemaRegistryUrl();
```
We can't remove it since `SchemaRegistryClient` doesn't return `schemaRegistryUrl`, do it?

iemejia replied:
`registryUrl` is an input parameter, not an output one. The user already knows the URL because he is building the client so that side is covered, and KafkaIO does not care about the `registryUrl` because it already has the built client. Or am I missing something?

aromanenko-dev replied:
We need to set Kafka config option "schema.registry.url" with this value and we do this in `withCSRClientProvider(CSRClientProvider)`. Unfortunately, `SchemaRegistryClient` doesn't have an API method to return `schemaRegistryUrl`, so we use the one stored in `CSRClientProvider`. What is an issue to have `String getSchemaRegistryUrl()` there?

iemejia replied:
Oh I see my mistake, sorry. Ugly `Map based APIs`. Ok then, but to the history I want to say that this is information leaking ! :stuck_out_tongue:

aromanenko-dev commented:
Run Java KafkaIO Performance Test

iemejia commented:
Merged manually to squash the review commits.

suztomo commented:
@aromanenko-dev @iemejia "SQL Postcommit" started failing after this commit. Would you check whether the failure is related to this PR?

![image](https://user-images.githubusercontent.com/28604/73775512-05e5e980-4754-11ea-8908-2e1b032d7090.png)

https://builds.apache.org/job/beam_PostCommit_SQL/3924/

```
org.apache.beam.sdk.extensions.sql.meta.provider.datacatalog.DataCatalogGCSIT.testReadFromGCS


java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
```

udim commented:
Opened https://issues.apache.org/jira/browse/BEAM-9251 for breakage of updateOfflineRepository

iemejia commented:
Thanks a lot @suztomo and @udim for taking care of this (time zones did not help for us to look),  a pity we do not have a check to detect such cases earlier.

suztomo commented:
I don’t know udim’s problem is related to my finding. SQL Postcommit is still red:
https://builds.apache.org/job/beam_PostCommit_SQL/

aromanenko-dev commented:
@suztomo Thanks, yes, it's still red. Do you know how to run this test locally?

suztomo commented:
No, I don't. Let me send an email to dev mailing list.

https://gist.github.com/suztomo/43de4ff1f0458e801fef1a16a28da301

aromanenko-dev commented:
@suztomo Thanks. I don't see a direct relation between this failing test (DataCatalogGCSIT) and KafkaIO. Only guess for now is that this PR could bring some transitive deps and it fails somehow because of this....
Btw, the full stacktrace when I run it locally:
https://gist.github.com/aromanenko-dev/0916b6d4f618a2c1b4f7b8ac5bf320ad
However, it's a different error that we have on Jenkins. Unfortunately, I can't see the details of Jenkins job since it's published into local file.

suztomo commented:
https://issues.apache.org/jira/browse/BEAM-9253

suztomo commented:
For BEAM-9253, Brian observes the data in the backend has changed on Feb 4th, and thus this PR is unrelated. Sorry for confusion.

aromanenko-dev commented:
@suztomo Thank you for taking care of this.
